gmso/external/convert_parmed.py

```diff
@@ -339,146 +386,296 @@ def _angle_types_from_pmd(structure, angle_types_member_map=None):
+            "RyckaertBellemansTorsionPotential"
+        ]
+        name = ryckaert_bellemans_torsion_potential.name
+        expression = ryckaert_bellemans_torsion_potential.expression
+        variables = ryckaert_bellemans_torsion_potential.independent_variables
 
         top_dihedraltype = gmso.DihedralType(
+            name=name,
             parameters=dihedral_params,
-            expression="c0 * cos(phi)**0 + c1 * cos(phi)**1 + "
-            + "c2 * cos(phi)**2 + c3 * cos(phi)**3 + c4 * cos(phi)**4 + "
-            + "c5 * cos(phi)**5",
-            independent_variables="phi",
-            member_types=member_types,
+            expression=expression,
+            independent_variables=variables,
+            member_types=dihedraltype[6],
         )
-        pmd_top_dihedraltypes[id(dihedraltype)] = top_dihedraltype
-    return pmd_top_dihedraltypes
+        pmd_top_rbtorsiontypes[dihedraltype] = top_dihedraltype
+    return pmd_top_rbtorsiontypes
 
 
-def _improper_types_from_pmd(structure, improper_types_member_map=None):
-    """Convert ParmEd improper types to GMSO ImproperType.
+def _improper_types_periodic_from_pmd(structure):
+    """Convert ParmEd DihedralTypes to GMSO improperType.
 
     This function take in a Parmed Structure, iterate through its
-    improper_types and dihedral_types with the `improper=True` flag,
-    create a corresponding GMSO.ImproperType, and finally return
-    a dictionary containing all pairs of pmd.ImproperType
-    (or pmd.DihedralType) and GMSO.ImproperType
-
+    dihedral_types with the improper flag, create a corresponding
+    GMSO.improperType, and finally return a dictionary containing all
+    pairs of pmd.impropertype and GMSO.improperType
     Parameters
     ----------
     structure: pmd.Structure
         Parmed Structure that needed to be converted.
-    improper_types_member_map: optional, dict, default=None
-        The member types (atomtype string) for each atom associated with the improper_types the structure
 
     Returns
     -------
     pmd_top_impropertypes : dict
-        A dictionary linking a pmd.ImproperType or pmd.DihedralType
-        object to its corresponding GMSO.ImproperType object.
+        A dictionary linking a pmd.improperType
+        object to its corresponding GMSO.improperType object.
     """
     pmd_top_impropertypes = dict()
-    improper_types_member_map = _assert_dict(
-        improper_types_member_map, "improper_types_member_map"
+    improper_dihedralsList = [
+        dihedral for dihedral in structure.dihedrals if dihedral.improper
+    ]
+    if len(improper_dihedralsList) == 0 or not structure.dihedral_types:
+        return pmd_top_impropertypes
+    unique_improper_types = list(
+        dict.fromkeys(
+            [
+                (
+                    improper.type.phi_k,
+                    improper.type.phase,
+                    improper.type.per,
+                    (atomtype_namegetter(improper)),
```

Comment:
Do you think we need to sort the members of improper (last 3 members)? since we are doing it for angle and dihedral.

Reply:
I would say yes, but I think I need to clearly understand what the way to do that is. We have so few cases for real impropers, and I'm not certain pmd.Impropers is sorted the same as pmd.dihedrals with improper=True. I think I need to figure those cases out.

Reply:
Addressed in [408ef1b](https://github.com/mosdef-hub/gmso/pull/721/commits/408ef1ba5b6b613ad3f9f541d395a4a99daaa1e5). Impropers are sorted by the index of the 2-4 sites, and the central atom is always listed first.